### PR TITLE
sphinx-build: Move code out of 'sphinx.__init__'

### PR DESCRIPTION
**Pull request by stephenfin**
Subject: sphinx-build: Move code out of `sphinx.__init__`

### Feature or Bugfix
- Refactor

### Purpose

We have multiple executables in tree and, while 'sphinx-build' is arguably the most important of these, there's no reason its importance should warrant inclusion at the package level.

Move this code to its own module, 'sphinx.build'. This name might be a bit disingenuous at present, given the availability of 'make-mode' here too, but that's an artifact of the current executable design and can be cleaned up later.

To avoid breaking packages that are using this feature directly, aliases for the old 'main' method are included. This is based on what [Django does](https://github.com/django/django/blob/1.11/django/test/runner.py#L688-L695) and, like Django, will allow us to safely remove the old modules in Sphinx 2.0.

### Detail
- N/A

### Relates
- N/A

### CODE REVIEW

**Comment by tk0miya** [comment]
@shimizukawa could you check this please.
I know you'd like to refactor command line interfaces on Sphinx-2.0.
I'm afraid this conflicts with the idea.

**Comment by stephenfin** [comment]
> @shimizukawa could you check this please.
> I know you'd like to refactor command line interfaces on Sphinx-2.0.
> I'm afraid this conflicts with the idea.

I'd like a little more info on this too. I've addressed #3259 locally, but have not pushed it because I'm waiting on #3929 to merge. Would be good to verify that I haven't wasted my time on this :smile:

**Comment by tk0miya** [comment]
I heard that he would like to restruct command line interface of Sphinx.
His idea is all commands (`sphinx-*`) will be integrated to sub-commands of `sphinx` command like following:
```
$ sphinx quickstart   (sphinx init?)
$ sphinx build
$ sphinx apidoc
# and so on.
```

Of course, we should keep current commands to keep compatibility for a while.
I don't know the change is important or not, but I feel good.

**Comment by stephenfin** [comment]
> I heard that he would like to restruct command line interface of Sphinx.
> His idea is all commands (sphinx-*) will be integrated to sub-commands of sphinx command like following:
> ```
> $ sphinx quickstart   (sphinx init?)
> $ sphinx build
> $ sphinx apidoc
> # and so on.
> ```

Ah, that makes sense. FWIW, it would be good to look at something like [`cliff`](https://docs.openstack.org/cliff/latest/) for this, which would allow plugins for folks that want to add other `sphinx` sub-commands. I'd be happy to look at this is @shimizukawa hadn't done anything yet.
_However_, while we definitely should wait for @shimizukawa's input, I can't see how this would have any ill-effects for that goal. If anything, it would make it easier, as it would allow us to include place the `sphinx` command stuff in `__init__.py` and call into `sphinx.build` for sub-commands.

**Comment by shimizukawa** [comment]
Actually, I have not done anything, so this change will not conflict with any ideas (I have not thought of anything yet!)

**Review by shimizukawa — APPROVED** [review]
LGTM!

File: sphinx/__init__.py [hunk 1]
```diff
@@ -63,62 +60,37 @@
...
-    """Sphinx build "make mode" entry."""
-    from sphinx import make_mode
-    return make_mode.run_make_mode(argv[1:])  # type: ignore
+def build_main(*args, **kwargs):
+    warnings.warn(
+        '`sphinx.build_main()` has moved to `sphinx.build.build_main()`.',
+        RemovedInSphinx20Warning,
+        stacklevel=2,
+    )
+    build.make_main(*args, **kwargs)
```
Comment (stephenfin) [inline sphinx/__init__.py]: Oops - this should be `build_main`. I'll rework

**Comment by stephenfin** [comment]
I fixed a typo. The patch is otherwise unchanged.

**Comment by tk0miya** [comment]
OMG, I noticed `sphinx/build.py` prevents my development. It blocks bash-completion for `sphinx/builders/`. I feel that is big problem to me. So I'd like to rename it to another naming...
How about `sphinx.cmdline.build` or `sphinx.commands.build`?

**Comment by stephenfin** [comment]
I'll do that now. Spotted some other cleanups that I could do in the interim

**Comment by stephenfin** [comment]
OK, I've moved this to `sphinx.cmd.build` instead. I haven't moved `sphinx.cmdline` and `sphinx.make_mode`, even though I wanted to, because I feel make mode could do with some work first. I'm going to wait for #4097 to merge first before doing any major surgery on this.

**Comment by tk0miya** [comment]
Thanks!

**Pull request merged**